### PR TITLE
feat: add helpers for encoding and signing l2deposit data

### PR DESCRIPTION
**Pull request by oneleo**
# feat: add helpers for encoding and signing l2deposit data
1. Add the ABI and contract file for L2Deposit.
2. Add the types for L2Deposit data.
3. Add the helpers for encoding L2Deposit data.
4. Add the helpers for signing L2Deposit data.
5. Add the addresses for the L2Deposit contract on the Goerli testnet.
6. Add the tests and utilities for the L2Deposit contract on the Goerli testnet.
7. Modify the .env.example, .gitignore, and build.yaml files.

## How to verify this pull request?
- Clone and build repo
```shell
% git clone git@github.com:consenlabs/tokenlon-contracts-lib-js.git
% code tokenlon-contracts-lib-js/
% git checkout LABS-1180/L2Deposit
% yarn install && yarn run build && yarn run compile-sol
```

- Edit the .env file
```shell
% cp .env.example .env && code .env

### Edit the .env file, 
### ------------------------------
CHAIN_ID="5" # e.g., 1 (mainnet), 5 (goerli), 42161 (arbitrum)
FORK_NODE_RPC_URL="https://eth-goerli.alchemyapi.io/v2/<YOUR API KEY>"
FORK_BLOCK_NUMBER="8185185"
### ------------------------------
```
> :warning: Please replace your API key with the FORK_NODE_RPC_URL variable

- Verify the L2Deposit library
``` shell
% yarn run test
```

- The result

<details>
  <summary>Output: (Please click here to expand)</summary>
    <pre><code>
yarn run v1.22.19
$ hardhat test

  L2Deposit
    Deposit on the Arbitrum bridge
      ✔ Should sign and encode a valid deposit for an EOA to use on the Arbitrum bridge. (844ms)
      ✔ Should sign and encode a valid deposit for an ERC1271 wallet to use on the Arbitrum bridge. (949ms)
      ✔ Should sign and encode a valid deposit for an ERC1271 wallet using ETHSign to use on the Arbitrum bridge. (121ms)
    Deposit on the Optimism bridge
      ✔ Should sign and encode a valid deposit for an EOA to use on the Optimism bridge. (349ms)
      ✔ Should sign and encode a valid deposit for an ERC1271 wallet to use on the Optimism bridge. (343ms)
      ✔ Should sign and encode a valid deposit for an ERC1271 wallet using ETHSign to use on the Optimism bridge. (94ms)

  6 passing (3s)

✨  Done in 8.27s.
    </code></pre>
</details>

Thank you very much!

### CODE REVIEW

File: test/v5/L2Deposit.ts [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+        describe("Deposit on the Optimism bridge", () => {
+            // Test for the EOA
+            it(`Should sign and encode a valid deposit for an EOA to use on the Optimism bridge.`, async () => {
+                const deposit = {
+                    // Override the default deposit data for the Optimism bridge
+                    ...defaultDeposit,
+                    l2Identifier: L2Identifier.Optimism,
+                    l1TokenAddr: token.USDT.address,
+                    l2TokenAddr: zeroAddress, // It does not matter for Optimism bridge
```
Comment (NIC619) [inline test/v5/L2Deposit.ts:211]: Maybe we should use an actual address to prevent people reading the tests for reference thinking that they can just fill in zero address for Optimism deposit.

File: network/goerli.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    Curve3Pool: "",
+
+    // L2 Bridge
+    ArbitrumL1GatewayRouter: "0x4c7708168395aEa569453Fc36862D2ffcDaC588c",
+    OptimismL1StandardBridge: "0x636Af16bf2f682dD3109e60102b8E1A089FedAa8",
+
+    // L2 Token
+    DAIForArbitrumBridge: "0x8411120Df646D6c6DA15193Ebe9E436c1c3a5222",
+    USDCForArbitrumBridge: "0x8FB1E3fC51F3b789dED7557E680551d93Ea9d892",
+    USDTForArbitrumBridge: "0xB401e876346B3C77DD51781Efba5223d2F1e6697",
```
[the remark this answers is not in the document]
Comment (oneleo) [reply network/goerli.ts:35]: No, but they are on the Arbitrum Goerli network.
DAI：[https://goerli.arbiscan.io/address/0x8411120Df646D6c6DA15193Ebe9E436c1c3a5222](https://goerli.arbiscan.io/address/0x8411120Df646D6c6DA15193Ebe9E436c1c3a5222)
USDC：[https://goerli.arbiscan.io/address/0x8FB1E3fC51F3b789dED7557E680551d93Ea9d892](https://goerli.arbiscan.io/address/0x8FB1E3fC51F3b789dED7557E680551d93Ea9d892)
USDT：[https://goerli.arbiscan.io/address/0xB401e876346B3C77DD51781Efba5223d2F1e6697](https://goerli.arbiscan.io/address/0xB401e876346B3C77DD51781Efba5223d2F1e6697)

They are used as the [parameters](https://github.com/consenlabs/tokenlon-contracts-lib-js/blob/LABS-1180/L2Deposit/test/v5/L2Deposit.ts#L64) for the Arbitrum bridge.
We can also obtain L2 tokens addresses from the 'calculateL2TokenAddress' function of the [Arbitrum L1 gateway contract on Goerli](https://goerli.etherscan.io/address/0x715D99480b77A8d9D603638e593a539E21345FdF#readProxyContract).

P.S.: The Goerli network addresses of the Arbitrum protocol: [https://developer.arbitrum.io/useful-addresses](https://developer.arbitrum.io/useful-addresses)

File: network/arbitrum.ts [hunk 3]
```diff
@@ -23,4 +24,13 @@ export default {
 
     // Curve
     Curve3Pool: "",
+
+    // L2 Bridge
+    ArbitrumL1GatewayRouter: "",
```
[the remark this answers is not in the document]
Comment (oneleo) [reply network/arbitrum.ts:29]: Yes, but these network files (arbitrum.ts, goerli.ts, mainnet.ts) all follow the same format as the [index.ts](https://github.com/consenlabs/tokenlon-contracts-lib-js/blob/LABS-1180/L2Deposit/network/index.ts#L19-L56) file,
which allows us to easily switch between them by modifying the CHAIN_ID parameter in the [.env](https://github.com/consenlabs/tokenlon-contracts-lib-js/blob/LABS-1180/L2Deposit/.env.example#L1) file.

File: test/utils/context.ts [hunk 4]
```diff
@@ -27,13 +28,17 @@ export type Context = {
         USDC: IERC20
         USDT: IERC20
         WETH: IERC20
+        DAIForArbitrumBridge: IERC20
+        USDCForArbitrumBridge: IERC20
+        USDTForArbitrumBridge: IERC20
```
Comment (pilagod) [inline test/utils/context.ts:33]: Maybe we need to come up a new way to organize the addresses since it starts to diversify between Mainnet and Arbitrum. We could think more and do the refinement in another PR.

File: test/v5/L2Deposit.ts [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+                        wallet.user.address,
+                        arbitrumGasData,
+                    ),
+                }
+                // Obtain a specific amount of tokens and approve their to be used by the AllowanceTarget contract
+                await dealTokenAndApprove(
+                    wallet.user,
+                    tokenlon.AllowanceTarget,
+                    deposit.l1TokenAddr,
+                    defaultDeposit.amount,
```
Comment (pilagod) [inline test/v5/L2Deposit.ts:75]: Use the deposit constructed in test instead of the default (I found other tests have the same issue, can be updated all together.)

File: test/v5/L2Deposit.ts [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+
+        // Calculate the value of the call as gas for the Arbitrum bridge
+        function calculateArbitrumCallValue(data: L2ArbitrumDepositData) {
+            const maxSubmissionCost = BigNumber.from(data.maxSubmissionCost)
+            const maxGas = BigNumber.from(data.maxGas)
+            return maxSubmissionCost.add(maxGas.mul(data.gasPriceBid))
+        }
+
+        // Verify that the event data are equal to the deposit data parameters
+        function assertEvent(receipt: ContractReceipt, deposit: L2Deposit) {
```
Comment (pilagod) [inline test/v5/L2Deposit.ts:344]: Maybe could rename it to `assertDepositedEvent` for clear.

File: test/v5/L2Deposit.ts (hunk 5 above)
Comment (oneleo) [reply test/v5/L2Deposit.ts:75]: No problem!
But I didn't find the same issue in any other test files except the [LimitOrder.ts](https://github.com/consenlabs/tokenlon-contracts-lib-js/blob/LABS-1180/L2Deposit/test/v5/LimitOrder.ts#L60-L90) file.
The approved amount parameters in this LimitOrder file are being used before the order constructed are declared, right?

**Review by pilagod — APPROVED** [review]

**Review by NIC619 — APPROVED** [review]

File: src/v5/encoding.ts [hunk 7]
```diff
@@ -160,8 +163,42 @@ export class EncodingHelper {
...
+                data.depositSig,
+            ],
+        ])
+    }
+
     /* Vendor */
 
+    public encodeL2ArbitrumDepositData(user: string, data: L2ArbitrumDepositData): string {
+        return ethers.utils.defaultAbiCoder.encode(
+            ["address", "uint256", "uint256", "uint256"],
```
Comment (charlesjhongc) [inline src/v5/encoding.ts:193]: It's better to declare the abi as `const` at the beginning.

File: test/v5/L2Deposit.ts [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+                await dealTokenAndApprove(
+                    wallet.user,
+                    tokenlon.AllowanceTarget,
+                    deposit.l1TokenAddr,
+                    deposit.amount,
+                    {
+                        walletContract: erc1271Wallet,
+                    },
+                )
+                // Add "\x19\x01" string to the data to be signed in order to avoid the risk of eth_sign.
```
[the remark this answers is not in the document]
Comment (oneleo) [reply test/v5/L2Deposit.ts:171]: The [encode()](https://github.com/ethers-io/ethers.js/blob/master/packages/hash/src.ts/typed-data.ts#L392-L398) function of ether.js repo is used by the [getEIP712Digest()](https://github.com/consenlabs/tokenlon-contracts-lib-js/blob/LABS-1180/L2Deposit/src/signing/signing.ts#L42-L46) and [getL2DepositEIP712Digest()](https://github.com/consenlabs/tokenlon-contracts-lib-js/blob/LABS-1180/L2Deposit/src/v5/signing.ts#L259) function to add the "0x1901" prefix.

File: src/v5/encoding.ts (hunk 7 above)
Comment (oneleo) [reply src/v5/encoding.ts:193]: I tried to add the `const` keyword at the beginning like below, but it shows an error saying that: `A class member cannot have the 'const' keyword.`

```
...
public const encodeL2ArbitrumDepositData(user: string, data: L2ArbitrumDepositData): string {
...
```

Comment (oneleo) [reply src/v5/encoding.ts:193]: I found that it can be declared as follows:
```
        const coder = ethers.utils.defaultAbiCoder
        return coder.encode(
        ...
```
I will try to replace it in my next commit, thanks!

File: test/v5/L2Deposit.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (charlesjhongc) [reply test/v5/L2Deposit.ts:171]: `Get the EIP-712 digest` is not equal to `add the \x19\x01` prefix. Check the function implementation.

File: src/v5/encoding.ts (hunk 7 above)
Comment (charlesjhongc) [reply src/v5/encoding.ts:197]: The abi part is `["uint8", "bytes"]` for example, not the coder.

Comment (oneleo) [reply src/v5/encoding.ts:197]: But how I can apply it here with the `const` keyword...
Does it mean that as below..?
```typescript
    public encodeL2ArbitrumDepositData(user: string, data: L2ArbitrumDepositData): string {
        const abi = ["address", "uint256", "uint256", "uint256"]
        return ethers.utils.defaultAbiCoder.encode(abi, [
            user,
            data.maxSubmissionCost,
            data.maxGas,
            data.gasPriceBid,
        ])
    }
```

**Review by charlesjhongc — APPROVED** [review]
